api/request_test.go

```diff
@@ -0,0 +1,224 @@
+package api
+
+import (
+	"reflect"
+	"testing"
+)
+
+func newSliceRequest(
+	vds       string,
+	sas       string,
+	direction string,
+	lineno    int,
+) SliceRequest {
+	return SliceRequest{
+		BaseRequest: BaseRequest{
+			Vds: vds,
+			Sas: sas,
+		},
+		Direction: direction,
+		Lineno:    &lineno,
+	}
+}
+
+func newFenceRequest(
+	vds              string,
+	sas              string,
+	coordinateSystem string,
+	coordinates      [][]float32,
+	interpolation    string,
+) FenceRequest {
+	return FenceRequest{
+		BaseRequest: BaseRequest{
+			Vds: vds,
+			Sas: sas,
+		},
+		CoordinateSystem: coordinateSystem,
+		Coordinates:      coordinates,
+		Interpolation:    interpolation,
+	}
+}
+
+func TestHashDoesNotAlterSliceRequest(t *testing.T) {
+	request := newSliceRequest("some-path", "some-sas", "inline", 9961)
+
+	before := request
+	_, err := request.Hash()
+	if err != nil {
+		t.Fatalf("Failed to compute hash, err: %v", err)
+	}
+
+	if !reflect.DeepEqual(before, request) {
+		msg := "Hash() changed object, before: %v, after: %v'"
+		t.Fatalf( msg, before, request)
+	}
+}
+
+func TestHashDoesNotAlterFenceRequest(t *testing.T) {
+	request := newFenceRequest(
+		"some-path",
+		"some-sas",
+		"ij",
+		[][]float32{{1,2}, {2,3}},
+		"linear",
+	)
+
+	before := request
+	_, err := request.Hash()
+	if err != nil {
+		t.Fatalf("Failed to compute hash, err: %v", err)
+	}
+
+	if !reflect.DeepEqual(before, request) {
+		msg := "Hash() changed object, before: %v, after: %v'"
+		t.Fatalf( msg, before, request)
+	}
+}
```

Comment:
This is one of the problems with go's error handling. I want to test 2), which is what the test name tries to communicate, but I have to do the call to hash() in order to do my test. I can discard the error from cache to better communicate that this is not what I'm testing here. But should that call fail, I wouldn't know and my actually test might or might not fail too. But now I don't know why it failed. Hence I believe it's better to explicitly verify that we are at a valid and well-defined state when we do our actual test. Because then we now that the test failed, not the setup to the test